File: app/src/main/java/com/example/adam/myapplication/ui/doctor/create/DoctorTaskFragment.java
package com.example.adam.myapplication.ui.doctor.create;


import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.TimePicker;

import com.example.adam.myapplication.R;
import com.example.adam.myapplication.data.objects.Doctor;
import com.example.adam.myapplication.ui.doctor.dialogs.ChooseDoctorDialog;
import com.example.adam.myapplication.utils.Formatter;
import com.example.adam.myapplication.utils.Picker;

import net.cachapa.expandablelayout.ExpandableLayout;

import java.text.Normalizer;
import java.util.Calendar;

public class DoctorTaskFragment extends Fragment {

    private View view;
    private DoctorTaskViewModel viewModel;

    private TextView doctorText;
    private TextView timeText;
    private TextView dateText;
    private Switch repeatSwitch;
    private ExpandableLayout expandableDateLayout;
    private TextView endDateText;
    private TextView moreLabel;
    private ExpandableLayout moreExpandable;
    private EditText noteText;
    private Button submitButton;

    public DoctorTaskFragment() {

    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_doctor_task, container, false);
        viewModel = ViewModelProviders.of(this).get(DoctorTaskViewModel.class);

        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        getComponents();
        setObservers();
        setListeners();
    }

    private void getComponents() {
        doctorText = view.findViewById(R.id.name_text);
        timeText = view.findViewById(R.id.time_text);
        dateText = view.findViewById(R.id.date_text);
        repeatSwitch = view.findViewById(R.id.switch_repeat);
        expandableDateLayout = view.findViewById(R.id.date_expandable);
        endDateText = view.findViewById(R.id.date_end_text);
        moreLabel = view.findViewById(R.id.more_label);
        moreExpandable = view.findViewById(R.id.more_expandable);
        noteText = view.findViewById(R.id.note_input);
        submitButton = view.findViewById(R.id.submit_button);
    }

    private void setObservers() {
        viewModel.getDoctor().observe(this, doctorObserver);
        viewModel.getDate().observe(this, dateObserver);
        viewModel.getEndDate().observe(this, endDateObserver);
    }

    private Observer<Doctor> doctorObserver = new Observer<Doctor>() {
        @Override
        public void onChanged(@Nullable Doctor doctor) {
            if (doctor != null)
                doctorText.setText(doctor.getSpecialization());
        }
    };

    private Observer<Calendar> dateObserver = new Observer<Calendar>() {
        @Override
        public void onChanged(@Nullable Calendar calendar) {
            if (calendar != null) {
                dateText.setText(Formatter.getDateString(calendar));
                timeText.setText(Formatter.getTimeString(calendar));
            }
        }
    };

    private Observer<Calendar> endDateObserver = new Observer<Calendar>() {
        @Override
        public void onChanged(@Nullable Calendar calendar) {
            if (calendar != null) {
                endDateText.setText(Formatter.getDateString(calendar));
            }
        }
    };

    private void setListeners() {
        doctorText.setOnClickListener(doctorListener);
        timeText.setOnClickListener(timeListener);
        dateText.setOnClickListener(dateListener);
        endDateText.setOnClickListener(endDateListener);
        repeatSwitch.setOnCheckedChangeListener(switchListener);
        moreLabel.setOnClickListener(moreListener);
        submitButton.setOnClickListener(submitListener);
    }

    private View.OnClickListener doctorListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {

        }
    };

    private View.OnClickListener dateListener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            Picker.showDatePicker(getActivity(), new DatePickerDialog.OnDateSetListener() {
                @Override
                public void onDateSet(DatePicker view, int year, int month, int day) {
                    viewModel.dateSet(year, month, day);
                }
            });
        }
    };

    private View.OnClickListener timeListener
            = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            Picker.showTimePicker(getActivity(), new TimePickerDialog.OnTimeSetListener() {
                @Override
                public void onTimeSet(TimePicker view, int hour, int minute) {
                    viewModel.timeSet(hour, minute);
                }
            });
        }
    };

    private CompoundButton.OnCheckedChangeListener switchListener = new CompoundButton.OnCheckedChangeListener() {
        @Override
        public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
            if (b)
                expandableDateLayout.expand();
            else
                expandableDateLayout.collapse();
        }
    };

    private View.OnClickListener endDateListener
            = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            Picker.showDatePicker(getActivity(), new DatePickerDialog.OnDateSetListener() {
                @Override
                public void onDateSet(DatePicker view, int year, int month, int day) {
                    viewModel.endDateSet(year, month, day);
                }
            });
        }
    };

    private View.OnClickListener moreListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            toggleMoreExpandable();
        }

        private void toggleMoreExpandable() {
            if (moreExpandable.isExpanded())
                moreExpandable.collapse();
            else
                moreExpandable.expand();
        }
    };

    private View.OnClickListener submitListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            viewModel.noteSet(noteText.getText().toString());
            viewModel.insertTask();
        }
    };
}

File: app/src/main/java/com/example/adam/myapplication/ui/doctor/create/DoctorTaskFragment.kt
package com.example.adam.myapplication.ui.doctor.create

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.example.adam.myapplication.R
import com.example.adam.myapplication.data.objects.Doctor
import com.example.adam.myapplication.ui.doctor.dialogs.ChooseDoctorDialog
import com.example.adam.myapplication.ui.doctor.dialogs.NewDoctorDialog
import com.example.adam.myapplication.ui.main.MainActivity
import com.example.adam.myapplication.utils.Formatter
import com.example.adam.myapplication.utils.Picker
import kotlinx.android.synthetic.main.fragment_doctor_task.*
import java.util.*

class DoctorTaskFragment : Fragment() {

    private lateinit var layout: View
    private lateinit var viewModel: DoctorTaskViewModel

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        layout = inflater.inflate(R.layout.fragment_doctor_task, container, false)
        viewModel = ViewModelProviders.of(this).get(DoctorTaskViewModel::class.java)

        return layout
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setObservers()
        setListeners()
    }

    private fun setObservers() {
        setDoctorObserver()
        setDateObserver()
        setEndDateObserver()
        setNoteObserver()
        setErrorObserver()
        setTaskCreatedObserver()
    }

    private fun setDoctorObserver() {
        viewModel.doctor.observe(this, Observer<Doctor> { doctor ->
            name_text.text = doctor?.specialization ?: ""
        })
    }

    private fun setDateObserver() {
        viewModel.date.observe(this, Observer<Calendar> { calendar ->
            date_text.text = Formatter.getDateString(calendar)
            time_text.text = Formatter.getTimeString(calendar)
        })
    }

    private fun setEndDateObserver() {
        viewModel.endDate.observe(this, Observer<Calendar> { calendar ->
            date_end_text.text = Formatter.getDateString(calendar)
        })
    }

    private fun setNoteObserver() {
        viewModel.note.observe(this, Observer { note ->
            if (note != null)
                note_input.setText(note)
        })
    }

    private fun setErrorObserver() {
        viewModel.error.observe(this, Observer { error ->
            showError(error!!)
        })
    }

    private fun showError(error: String) {
        Toast.makeText(activity, error, Toast.LENGTH_LONG).show()
    }

    private fun setTaskCreatedObserver() {
        viewModel.task.observe(this, Observer { task ->
            if (task != null)
                (activity as MainActivity).goToBoard()
        })
    }

    private fun setListeners() {
        setDoctorListener()
        setDateListener()
        setTimeListener()
        setSwitchListener()
        setEndDateListener()
        setMoreListener()
        setSubmitListener()
    }

    private fun setDoctorListener() {
        name_text.setOnClickListener {
            showChooseDoctorDialog()
        }
    }

    private fun setDateListener() {
        date_text.setOnClickListener {
            Picker.showDatePicker(activity) { _, year, month, dayOfMonth ->
                viewModel.dateSet(year, month, dayOfMonth)
            }
        }
    }

    private fun setTimeListener() {
        time_text.setOnClickListener {
            Picker.showTimePicker(activity) { _, hour, minute ->
                viewModel.timeSet(hour, minute)
            }
        }
    }

    private fun setSwitchListener() {
        switch_repeat.setOnCheckedChangeListener { _, isChecked ->
            viewModel.isCyclic(isChecked)

            if (isChecked)
                date_expandable.expand()
            else
                date_expandable.collapse()
        }
    }

    private fun setEndDateListener() {
        date_end_text.setOnClickListener {
            Picker.showDatePicker(activity) { _, year, month, day ->
                viewModel.endDateSet(year, month, day)
            }
        }
    }

    private fun setMoreListener() {
        more_label.setOnClickListener {
            toggleMoreExpandable()
        }
    }

    private fun toggleMoreExpandable() {
        if (more_expandable.isExpanded)
            more_expandable.collapse()
        else
            more_expandable.expand()
    }

    private fun setSubmitListener() {
        submit_button!!.setOnClickListener {
            viewModel.noteSet(note_input.text.toString())
            viewModel.submitTask()
        }
    }

    private fun showChooseDoctorDialog() {
        val builder = ChooseDoctorDialog.Builder(this)

        builder.setItemClickListener { _, doctor ->
            viewModel.doctorChosen(doctor)
        }

        builder.showNewDoctorButton {
            showCreateDoctorDialog()
        }

        builder.create().show()
    }

    private fun showCreateDoctorDialog() {
        NewDoctorDialog.Builder(this)
                .setResultListener(newDoctorListener)
                .create()
                .show()
    }

    private val newDoctorListener = NewDoctorDialog.Builder.OnResultListener {
        status, doctor, _, error ->

        when (status) {
            NewDoctorDialog.SUCCESS -> viewModel.doctorChosen(doctor)
            NewDoctorDialog.FAILURE -> showError(error)
        }
    }
}
